src/KeepaliveConnector.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace React\HttpClient;
+
+use React\Promise\PromiseInterface;
+use React\Socket\ConnectionInterface;
+use React\Socket\ConnectorInterface;
+
+/**
+ * @internal
+ */
+class KeepaliveConnector implements ConnectorInterface
+{
+    private $directConnector;
+
+    /** @var PromiseInterface[] */
+    private $aliveConnections = array();
+
+    public function __construct(ConnectorInterface $directConnector)
+    {
+        $this->directConnector = $directConnector;
+    }
+
+    public function connect($uri)
+    {
+        return isset($this->aliveConnections[$uri])
+            ? $this->tryReuseAliveConnection($uri)
+            : $this->createNewConnection($uri);
+    }
+
+    public function handleConnectionClose($uri)
+    {
+        unset($this->aliveConnections[$uri]);
+    }
+
+    private function tryReuseAliveConnection($uri)
+    {
+        return $this->aliveConnections[$uri];
```

Comment:
there was a check for `ConnectionInterface#isWritable()` but then i realized `$this->aliveConnections` is actually an array of Promises.